### PR TITLE
Fix issue where `cy.wrap` does not respect timeout option

### LINKED ISSUE #5980: cy.wrap does not respect timeout
This example spec file fails with the error below:

```js
describe('Cypress wait', () => {
  it('wrap does not wait', () => {
    cy.wrap(
        new Promise((resolve) => setTimeout(resolve, 5000)), {timeout: 20000});
  });
});
```

```
 Cypress wait
    1) wrap does not wait


  0 passing (4s)
  1 failing

  1) Cypress wait wrap does not wait:
     Error: Cypress command timeout of '4000ms' exceeded.
      at http://localhost:57838/__cypress/runner/cypress_runner.js:103532:25
```

Shouldn't `cy.wrap` respect the `timeout` option? The test passes if I pass `--config defaultCommandTimeout=6000` on the command line, but I want to pass a specific timeout for this `cy.wrap` command.

### PR DESCRIPTION
**Pull request by panzarino**
- Closes #5980

### User facing changelog

- Adds proper functionality to the `timeout` option for `cy.wrap()`

### User facing changes
error message when promise passed to `cy.wrap(...)` times out

before (also does not respect `{timeout}` option):
image

after:
image

### Additional details

- ~~This option has been documented for a while now but actually has not been implemented.~~
- These changes update the timeout option to also include any promises passed to the `cy.wrap` command in addition to the assertions that follow.
- NOTE: **This is a breaking change.**
  - Previously, any promises passed as parameters to any `cy` command would be awaited before the function was called, and the function would be called with the result of the promise.
  - With these changes, the promises will be passed directly to the function without evaluation fist.
  - As the entire test suite still passes with these changes, it indicates that this change of functionality is fixing a bug rather than removing a feature.   

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?

### CODE REVIEW

**Comment by Bkucera** [comment]
wow, it looks to be an old api design decision : https://github.com/cypress-io/cypress/commit/1ca5bed33c7f8bff40ed917c8273c934b4cd7ceb#diff-c220d94ca4f17f07a8f6b1932cc3bcc1R18

since no tests fail, this looks unintended

**Comment by panzarino** [comment]
@Bkucera so playing around with some tests for a while, it appears that the timeout function does actually serve a purpose and was implemented correctly. The failing tests here demonstrate that. It sets a timeout on the assertions that follow the wrap function, so that when those assertions are retried it will fail once it times out. These tests are both failing/passing because sometimes it will time out due to assertions and throw an assertion error, and other times due to these changes with a Cypress error.

Therefore I think there can be two ways to proceed: either fix these changes so that the assertion error is thrown every time before the cypress timeout error, or remove these changes entirely and leave the functionality as is.

File: packages/driver/src/cy/commands/misc.js [hunk 1]
```diff
@@ -50,6 +58,13 @@ module.exports = (Commands, Cypress, cy) => {
           onRetry: resolveWrap,
         })
         .return(arg)
+        .timeout(options.timeout)
+        .catch(Promise.TimeoutError, () => {
+          $errUtils.throwErrByPath('wrap.timed_out', {
+            onFail: options._log,
+            args: { timeout: options.timeout },
+          })
+        })
```
Comment (Bkucera) [inline packages/driver/src/cy/commands/misc.js:67]: is this necessary? I believe cy.verifyUpcomingAssertions will await `arg` with the proper timeout, maybe that is causing the issue you describe

Comment (panzarino) [reply packages/driver/src/cy/commands/misc.js:67]: It doesn't look like that is the case, ran a couple tests locally without these changes and they aren't respecting the timeout in the way intended with these changes (exhibits the same behavior as before we removed `all()`).

Aside from that and assuming that the inconsistent errors are fixed, it looks like what this change is doing is actually modifying the way that the `timeout` option had already been getting processed which may or may not be desired.

**Review by kuceb — DISMISSED** [review]
approving the code, @jennifer-shehane can you review the user-facing-changes in the PR comment?

**Comment by jennifer-shehane** [comment]
Holding off on merging until I get clarification on this being a breaking change - whether it requires a major version release.

**Comment by jennifer-shehane** [comment]
@panzarino Will this also fix this issue? https://github.com/cypress-io/cypress/issues/1221

**Comment by panzarino** [comment]
@jennifer-shehane Unfortunately this will not fix that issue. I ran a couple tests as described in that issue and the behavior is still the same. According to the docs, there is a timeout option for `invoke`, which looks like it should describe the behavior requested by that issue. This fix is only for the timeout option of `wrap`, but I can look into fixing the implementation of the timeout option for `invoke` as a separate PR.

**Review by jennifer-shehane — DISMISSED** [review]
Pending review from @brian-mann to pair with @panzarino on this.

**Review by brian-mann — CHANGES_REQUESTED** [review]
- lets add a test that verifies only a single `cy.wrap()` command log is inserted when given a rejected promise as a parameter
- also when the promise times out verify only a single command log is inserted

**Review by brian-mann — APPROVED** [review]

**Pull request merged**